pkg/scheduler/scheduler.go

```diff
@@ -227,7 +227,59 @@ func (m *Scheduler) processAllocationReleaseByAllocationKey(allocationAsksToRele
     }
 }
 
+func (m *Scheduler) recoverExistingAllocations(existingAllocations []*si.Allocation, rmId string) {
+    // the recovering of existing allocations looks like a replay of the scheduling process,
+    // in general, a scheduling process takes following steps
+    //  1) add scheduling info to schedulers cache, including app, pending requests etc
+    //  2) Look order partition -> queue -> app
+    //  3) For pending asks of a app, scheduler runs scheduling logic to propose a node
+    //  4) Process the proposal generated in step 3, update data in cache
+    // the recovery is repeating all steps except step 2
+    for _, alloc := range existingAllocations {
+        log.Logger.Info("recovering allocations for app",
+            zap.String("applicationId", alloc.ApplicationId),
+            zap.String("nodeId", alloc.NodeId),
+            zap.String("queueName", alloc.QueueName),
+            zap.String("partition", alloc.Partition),
+            zap.String("allocationId", alloc.Uuid))
+
+        // add scheduling asks
+        schedulingAsk := ConvertFromAllocation(alloc, rmId)
+        if err := m.updateSchedulingRequest(schedulingAsk); err != nil {
```

Comment:
How do we know that the application is already added to the cache?
I don't see the application being added to the cache on recovery

Reply:
It relies on shim to provide info of existing applications. When shim is in recovery mode, it scans all existing apps, and simply send to scheduler-core as new applications, e.g

```
&si.UpdateRequest{
    ....,
    NewApplications: []*si.AddApplicationRequest{
			{
				ApplicationId: "app-1",
				QueueName:     "root.a",
				PartitionName: "",
			},
		},
}
```
when core gets these info, it adds application to cache just like a fresh submitted app.

Reply:
Can I propose the following then:
- Add to the design that we make sure the shim sends the application first (full scan).
- The queue for events must not have any new application events waiting when we start processing the allocations and  or node additions. If we do there is a big chance of failures.

Reply:
for your proposed points
> Add to the design that we make sure the shim sends the application first (full scan)

Yes, we'll make sure shim sends existing apps and allocations first. We'll do a full node-scan with node lister api. I'll add this to the design.

> The queue for events must not have any new application events waiting when we start processing the allocations and or node additions. If we don't there is a big chance of failures.

We will also ensure this from the shim layer.

Reply:
OK, will wait for the shim changes to confirm this
with that pre-requisite I am OK